pyro/distributions/diag_normal_mixture.py

```diff
@@ -0,0 +1,181 @@
+    def forward(ctx, locs, scales, component_logits, pis, which, noise_shape):
+        dim = scales.size(-1)
+        white = locs.new(noise_shape).normal_()
+        n_unsqueezes = locs.dim() - which.dim()
+        for _ in range(n_unsqueezes):
+            which = which.unsqueeze(-1)
+        which_expand = which.expand(tuple(which.shape[:-1] + (dim,)))
+        loc = torch.gather(locs, -2, which_expand).squeeze(-2)
+        sigma = torch.gather(scales, -2, which_expand).squeeze(-2)
+        z = loc + sigma * white
+        ctx.save_for_backward(z, scales, locs, component_logits, pis)
+        return z
+
+    @staticmethod
+    @once_differentiable
+    def backward(ctx, grad_output):
+
+        z, scales, locs, logits, pis = ctx.saved_tensors
+        dim = scales.size(-1)
+        K = logits.size(-1)
+        g = grad_output  # l b i
+        g = g.unsqueeze(-2)  # l b 1 i
+        batch_dims = locs.dim() - 2
+
+        locs_tilde = locs / scales  # b j i
+        sigma_0 = torch.min(scales, -2, keepdim=True)[0]  # b 1 i
+        z_shift = (z.unsqueeze(-2) - locs) / sigma_0  # l b j i
+        z_tilde = z.unsqueeze(-2) / scales - locs_tilde  # l b j i
+
+        mu_cd = locs.unsqueeze(-2) - locs.unsqueeze(-3)  # b c d i
+        mu_cd_norm = torch.pow(mu_cd, 2.0).sum(-1).sqrt()  # b c d
+        mu_cd /= mu_cd_norm.unsqueeze(-1)  # b c d i
+        diagonals = torch.LongTensor(range(K))
+        mu_cd[..., diagonals, diagonals, :] = 0.0
+
+        mu_ll_cd = (locs.unsqueeze(-2) * mu_cd).sum(-1)  # b c d
+        z_ll_cd = (z.unsqueeze(-2).unsqueeze(-2) * mu_cd).sum(-1)  # l b c d
+        z_perp_cd = z.unsqueeze(-2).unsqueeze(-2) - z_ll_cd.unsqueeze(-1) * mu_cd  # l b c d i
+        z_perp_cd_sqr = torch.pow(z_perp_cd, 2.0).sum(-1)  # l b c d
+
+        shift_indices = torch.LongTensor(range(dim)) - 1
+        shift_indices[0] = 0
+        z_shift_cumsum = torch.pow(z_shift, 2.0)
+        z_shift_cumsum = z_shift_cumsum.sum(-1, keepdim=True) - torch.cumsum(z_shift_cumsum, dim=-1)  # l b j i
+        z_tilde_cumsum = torch.cumsum(torch.pow(z_tilde, 2.0), dim=-1)  # l b j i
+        z_tilde_cumsum = torch.index_select(z_tilde_cumsum, -1, shift_indices)
+        z_tilde_cumsum[..., 0] = 0.0
+        r_sqr_ji = z_shift_cumsum + z_tilde_cumsum  # l b j i
+
+        log_scales = torch.log(scales)  # b j i
+        epsilons_sqr = torch.pow(z_tilde, 2.0)  # l b j i
+        log_qs = -0.5 * epsilons_sqr - 0.5 * np.log(2.0 * np.pi) - log_scales  # l b j i
+        log_q_j = log_qs.sum(-1, keepdim=True)  # l b j 1
+        q_j = torch.exp(log_q_j)  # l b j 1
+        q_tot = (pis * q_j.squeeze(-1)).sum(-1)  # l b
+        q_tot = q_tot.unsqueeze(-1)  # l b 1
+
+        root_two = math.sqrt(2.0)
+        shift_log_scales = log_scales[..., shift_indices]
+        shift_log_scales[..., 0] = 0.0
+        sigma_products = torch.cumsum(shift_log_scales, dim=-1).exp()  # b j i
+
+        reverse_indices = torch.LongTensor(range(dim - 1, -1, -1))
```

Comment:
See above for cuda-compatible `torch.arange` invocation.

Reply:
If `arange` didn't work, maybe `sigma_0.new_tensor(range(dim - 1, -1, -1), dtype=torch.long)` would work?